File: go/src/2020/july/add-binary.go
package july

func addBinary(a string, b string) string {
	if len(a) == 0 {
		return a
	}
	if len(b) == 0 {
		return b
	}
	if a[len(a)-1] == '1' && b[len(b)-1] == '1' {
		return addBinary(addBinary(a[0:len(a)-2], b[0:len(b)-2]), "1") + "0"
	}
	if a[len(a)-1]== '0' && b[len(b)-1] == '0' {
		return addBinary(a[0:len(a)-2], b[0:len(b)-2]) + "0"
	} else {
		return addBinary(a[0:len(a)-2], b[0:len(b)-2]) + "1"
	}
}

File: archive/java/src/main/java/cn/az/code/year2020/july/AddBinary.java
package cn.az.code.year2020.july;

import java.math.BigInteger;

/**
 * @author az
 * @since 07/19/20
 */
public class AddBinary {

    public static void main(String[] args) {
        AddBinary ab = new AddBinary();
        System.out.println(ab.addBinary("1011", "1011"));
    }

    public String addBinary(String a, String b) {
        return new BigInteger(a, 2).add(new BigInteger(b, 2)).toString(2);
    }

    public String _addBinary(String a, String b) {
        int i = a.length() - 1;
        int j = b.length() - 1;

        StringBuilder sb = new StringBuilder();
        int carry = 0;
        int sum = 0;
        while (i >= 0 && j >= 0) {
            sum = a.charAt(i--) - '0' + b.charAt(j--) - '0' + carry;
            sb.append(sum % 2);
            carry = sum / 2;
        }

        while (i >= 0) {
            sum = a.charAt(i--) - '0' + carry;
            sb.append(sum % 2);
            carry = sum / 2;
        }

        while (j >= 0) {
            sum = b.charAt(j--) - '0' + carry;
            sb.append(sum % 2);
            carry = sum / 2;
        }

        if (carry != 0) {
            sb.append(carry);
        }

        return sb.reverse().toString();
    }

    public String __addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int i = a.length() - 1, j = b.length() - 1, carry = 0;
        while (i >= 0 || j >= 0) {
            int sum = carry;
            if (j >= 0) sum += b.charAt(j--) - '0';
            if (i >= 0) sum += a.charAt(i--) - '0';
            sb.append(sum % 2);
            carry = sum / 2;
        }
        if (carry != 0) sb.append(carry);
        return sb.reverse().toString();
    }
}
